tensorflow/python/ops/numpy_ops/np_array_ops.py

```diff
@@ -560,6 +560,19 @@ def _reduce(tf_fn,
       tf_fn(input_tensor=a.data, axis=axis, keepdims=keepdims))
 
 
+@utils.np_doc('size')
+def size(x):
+  x = x.numpy() if isinstance(x, tf.Tensor) else x
```

Comment:
This currently won't handle graph tensors. How about something similar to below to handle those as well?

```suggestion
  x = asarray(x).data
  sz_tensor = array_ops.size(x)
  sz_value = np_utils.get_static_value(sz_tensor)
  if sz_value is not None:
    return int(sz_value)
  return np_utils.tensor_to_ndarray(sz_tensor)
```

Reply:
Thanks for the insight! Just to make sure, can we turn a graph Tensor directly into TF NumPy array?

Reply:
Yes! The ndarray class can wrap either graph or eager tensors.

Reply:
Thanks Akshay! I will commit this a bit later since I am currently dealing with some Neural Tangents issues.

Reply:
Unresolving this since it looks like the code still doesn't handle graph tensors.

Reply:
Hi Akshay @akshaym , thanks for pointing this out! I further revised the code in https://github.com/tensorflow/tensorflow/pull/42539/commits/fc5de5045086933fa7a657f0133b353a80b5d04a and now it should work fine.

Reply:
Hi Zhibo, the code as written still doesn't handle graph tensors. 

Graph tensors can have undefined shapes during graph construction (so for example the tensor can have a shape of "[None, None]")

In this situation [tnp.ndarray.shape](https://github.com/tensorflow/tensorflow/blob/a067c6723019ea08d3b4a1acf2cb1dbd42e33e09/tensorflow/python/ops/numpy_ops/np_arrays.py#L197) will return a tf.Tensor, which you cannot pass to np.prod.

Reply:
Thanks for mentioning this, Akshay. In your opinion, how should we handle undefined shapes ... raise `ValueError`?

Reply:
By the way, would you mind providing a concrete example on what a `None` shape Tensor construction is like? Thanks.

Reply:
The suggestion I gave above (in the first comment on this thread) should handle both graph and eager tensors. 

Here is a contrived example:
```
@tf.function(input_signature=[tf.TensorSpec(shape=None, dtype=tf.float32)])
def f(x):
  x = tnp.asarray(x)  # Wraps a graph tensor into a tnp.ndarray
  return tnp.size(x)  # This should fail since the shape is "None"
```

Reply:
Hi Akshay, thanks for the insight! I tried wrapping the `tf.TensorSpec(shape=None, dtype=tf.float32)` using `np.asarray`, but got the error `TypeError: data type not understood`.

Reply:
Hi @akshaym , I revised the code in https://github.com/tensorflow/tensorflow/pull/42539/commits/d70b16b8a54d15afc9b0fc33f0bbe93ba22ba456 based on your suggestion.